grub: Add `boot.loader.grub.extraGrubInstallArgs` option

nh2 opened the pull request:
###### TODO

* [x] Based upon #92205 merge that first
  * [x] and bump out WIP commits
* [x] Based upon #92122, merge that first

###### Motivation for this change

Useful for when you need to build grub modules into your grub kernel to get a working boot, as shown in the added example, but also for many other use cases where you want to customise the call to `grub-install`.

Related `grub-help` thread, whose members kindly helped out understanding the related issues: https://lists.gnu.org/archive/html/help-grub/2020-04/msg00004.html

Also fixes the NixOS `/boot/grub/state` file not being written atomically, and I added a couple comments in case some other people have to update this perl script in the future.

I've ensured backwards compatibility of the state file. @chpatrick has co-authored this.

I've tested this on my dedicated server, the one from https://discourse.nixos.org/t/howto-install-nixos-on-an-ovh-dedicated-server/3089/4?u=nh2.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

nh2 committed:
install-grub.pl: Refactor: Extract `getList()`

nh2 committed:
grub: Add `boot.loader.grub.extraGrubInstallArgs` option.

Useful for when you need to build grub modules into your grub kernel
to get a working boot, as shown in the added example.

To store this new value, we switch to more structural JSON approach.

Using one line per value to store in `/boot/grub/state` gets really messy when
the values are arrays, or even worse, can contain newlines (escaping would be
needed). Further, removing a value from the file would get extra messy
(empty lines we'd have to keep for backwards compatibility).

Thus, from now on we use JSON to store all values we'll need in the future.

nh2 committed:
install-grub.pl: Write state file atomically.

Other files were already written atomically, but not this one.

nh2 commented:
Side note: I think the line-based `/boot/grub/state` format in that perl script is suboptimal, as it has no real backwards compatibility, and cannot cleanly encode arbitrary data types (e.g. I had to use JSON to encode a list of strings). The `devices` list is encoded as comma-separated paths, which works in practice but is not fully technically correct because paths can have commas in them.

I think it would be better to switch the state file fully to a proper serialisation format, like JSON; that would also make the code cleaner and adding more fields easier. But that is for a another PR.

nh2 commented:
@GrahamcOfBorg test installer

nh2 commented:
Asking for review: @symphorien @emilazy @Mic92 @Kloenk because you reviewed a similar GRUB PR #85418, so maybe you can review this one as well :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/192

OmnipotentEntity commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -60,6 +60,7 @@ let
       inherit (efi) canTouchEfiVariables;
       inherit (cfg)
         version extraConfig extraPerEntryConfig extraEntries forceInstall useOSProber
+        extraGrubInstallArgs
```
This style is nice for keeping the diff clean.  But I would prefer that the lines were similar lengths, or if similar things were grouped together.

nh2 replied:
I need to optimise for a clean, minimal diff here, as I need to cherry-pick this into various projects / backports and need to avoid merge conflicts.

If we want to do extra grouping, I think it should be done in a separate PR.

OmnipotentEntity commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -236,6 +237,30 @@ in
 
+      extraGrubInstallArgs = mkOption {
+        default = [ ];
+        example = [ "--modules=nativedisk ahci pata part_gpt part_msdos diskfilter mdraid1x lvm ext2" ];
```
From this, it's not immediately clear to me whether separate arguments go in separate string entries, or if it's allowable to have several options together e.g.:

```[ "--boot-directory=/mnt/usb/bugbios --force --allow-floppy /dev/loop0" ]```

or must it be:

```[ "--boot-directory=/mnt/usb/bugbios" "--force" "--allow-floppy /dev/loop0" ]```

?

OmnipotentEntity replied:
Additionally, it's not clear if ordering is respected, this may be assumed by a user, especially in the first case, where he or she might want to do something like:

```{enableAhci ? true}:
  {
    achiStr = "" ++ lib.optionalString "ahci";
    extraGrubInstallArgs = [ "--modules=nativedisk" ahciStr "pata part_gpt part_msdos diskfilter mdraid1x lvm ext2" ]
  }
```

(Reading down, it looks like the arguments are just passed to the perl system command, so ordering is respected, and arguments can be combined and split.  So we are dealing with the first case.)

nh2 replied:
> whether separate arguments go in separate string entries, or if it's allowable to have several options together

Separate `grub-install` arugments go into separate string entries. [Example for passing modules this way.](
https://lists.gnu.org/archive/html/help-grub/2012-03/msg00085.html).

The perl script uses [`system PROGRAM LIST`](https://perldoc.perl.org/functions/system.html) to spawn `grub-install`, which explains

> If there is more than one argument in LIST, or if LIST is an array with more than one value, starts the program given by the first element of the list with arguments given by the rest of the list. If there is only one scalar argument, the argument is checked for shell metacharacters, ...

We pass a `LIST` to it, so the individual elements go straight to `execve()`. 

I will add this to `description`:

```
The list elements are passed directly as <literal>argv</literal>
arguments to the <literal>grub-install</literal> program, in order.
````

---

Just for extra clarity for anyone reading, this means that 

> `extraGrubInstallArgs = [ "--modules=nativedisk" ahciStr "pata part_gpt part_msdos diskfilter mdraid1x lvm ext2" ]`

from your example is illegal, because grub expects `--modules=space separated list of modules` to be a single argv argument.

OmnipotentEntity commented on nixos/modules/system/boot/loader/grub/install-grub.pl:
```diff
@@ -572,24 +586,26 @@ sub readGrubState {
+my @prevExtraGrubInstallArgs = @{decode_json($prevGrubState->extraGrubInstallArgs)};
 
 my $devicesDiffer = scalar (List::Compare->new( '-u', '-a', \@deviceTargets, \@prevDeviceTargets)->get_symmetric_difference());
+my $extraGrubInstallArgsDiffer = scalar (List::Compare->new( '-u', '-a', \@extraGrubInstallArgs, \@prevExtraGrubInstallArgs)->get_symmetric_difference());
```
This line seems to be the reason why json was used?  Is this robust though?  Will the following trigger a rebuild?

```
extraGrubInstallArgs = [ "--arg1=foo --arg2=bar" ];
prevExtraGrubInstallArgs = [ "--arg1=foo" "--arg2=bar" ];
```

nh2 replied:
> This line seems to be the reason why json was used?

No, JSON is used because the `/boot/grub/state` file remembers all arguments/parameters of the last successful `install-grub` run (to avoid unnecessary installation of e.g. the GRUB MBR on every `nixos-rebuild` run, which takes a couple seconds and would make no-op rebuilds slow), and the serialisation format that the state file used so far is line-based, which worked so far because there was only 1 word on each line. But as soon as you want to serialise something into it that is a list or can even contain newlines, it gets very messy and likely incorrect, so I use JSON instead and we'll use that going forward.

Your question motivated me to rework the the way this works, full details in https://github.com/NixOS/nixpkgs/pull/85895#issuecomment-653309518.

nh2 replied:
>  Is this robust though? Will the following trigger a rebuild?

Yes, because these are different (and the first line will be rejected by `grub-install`), see https://github.com/NixOS/nixpkgs/pull/85895#discussion_r449349168.

OmnipotentEntity reviewed (commented):
I saw your post on Discourse.  But I don't have a good way of testing this PR for functionality.  So instead, I thought I'd just give some general feedback.

(Thanks for the PR btw, it looks well done.)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/how-to-build-a-nixos-vm-with-nix-in-which-nixos-rebuild-is-a-no-op/7937/1

nh2 committed:
install-grub.pl: Add errno messages to all `or die` errors.

For example, turns the error

    cannot copy /nix/store/g24xsmmsz46hzi6whv7qwwn17myn3jfq-grub-2.04/share/grub/unicode.pf2 to /boot

into the more useful

    cannot copy /nix/store/g24xsmmsz46hzi6whv7qwwn17myn3jfq-grub-2.04/share/grub/unicode.pf2 to /boot: Read-only file system

nh2 commented:
Update: Now

* [ ] Based upon #92122, merge that first

And I've written a repo with a VM test that makes it trivial to test this change:

https://github.com/nh2/nixos-vm-examples

In particular: https://github.com/nh2/nixos-vm-examples/blob/0b29937fcb23424b95aa98121c94db02fd9b920d/grub-test-vm/configuration.nix

nh2 commented:
I have reworked the way JSON is used to store contents in `/boot/grub/state` to make it more general, and put an end to the line-based unstructured serialisation format.

Details:

The `/boot/grub/state` file remembers all arguments/parameters of the last successful `install-grub` run (to avoid unnecessary installation of e.g. the GRUB MBR on every `nixos-rebuild` run, which takes a couple seconds and would make no-op rebuilds slow), and the serialisation format that the state file used so far is line-based, which worked so far because there was only 1 word on each line. But as soon as you want to serialise something into it that is a list or can even contain newlines, it gets very messy and likely incorrect, so I use JSON instead and we'll use that going forward.

I reworked the the way this works (with my last force-push), so now it's not just a single JSON list for the new param I've added, but a JSON object on line 6 of the state file, which can take care of all further parameters that need to be added in the future, as structured data with automatic, correct escaping (as correct as you can make it with Perl; this contribution took a long time and way longer than this type of work should take in 2020).

Example of what the state file looks like now:

```
[root@nixos:~]# cat /boot/grub/state 
grub
2.04
no
/dev/vdb
/boot
{"extraGrubInstallArgs":[]}
```

So, further changes should go into the JSON object in the last line.

nh2 commented:
> I saw your post on Discourse. [...] So instead, I thought I'd just give some general feedback.

Thanks, much appreciated. I made some improvements based on your comments.

> But I don't have a good way of testing this PR for functionality.

https://github.com/NixOS/nixpkgs/pull/85895#issuecomment-653306643 should address that, but you do not need to test it unless you'd like to, as I've tested with that quite exhaustively now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-vm-examples-a-collection-of-examples-on-how-to-build-nixos-vms/7939/3

nh2 commented:
#92122 is merged, so this is ready to go too.

petabyteboy commented:
One of these very recent changes in the Grub module has caused the following error on every host I deployed:

```
Can't use an undefined value as an ARRAY reference at /nix/store/a32rgs0n0vnk95lkb8d7wcywd0svjb6k-install-grub.pl line 642, <FILE> line 5.
```

The state file contains the following: 

```
grub
2.04
no
/dev/sda
/boot

```

nh2 commented:
@petabyteboy That is bad, I had gotten some similar errors too while developing this but then I learned more Perl and thought I fixed all possible cases. I will look into it immediately; thanks for posting your state file.

nh2 commented:
@petabyteboy I think I have the fix, testing it now on your setup and with my VM tests from https://github.com/nh2/nixos-vm-examples.

nh2 commented:
@petabyteboy Could you try PR #92520?
